lib/vdsm/virt/thinp.py

```diff
@@ -214,13 +220,43 @@ def on_block_threshold(self, target, path, threshold, excess):
                 'Unknown drive %r for vm %s - ignored block threshold event',
                 drive_name, self._vm.id)
         else:
-            drive.on_block_threshold(path)
+            if drive.on_block_threshold(path):
+                self._extend_drive_soon(drive)
 
     def on_enospc(self, drive):
         """
         Called when a VM pauses because of ENOSPC error writing to drive.
         """
-        drive.on_enospc()
+        if drive.on_enospc():
+            self._extend_drive_soon(drive)
+
+    def _extend_drive_soon(self, drive):
+        if self._dispatch is None:
+            return
+        self._log.debug("Schedudling drive %s extension", drive.name)
+        try:
+            self._dispatch(
+                partial(self._extend_drive, drive),
+                timeout=config.getfloat("thinp", "extend_timeout"))
+        except exception.ResourceExhausted:
+            # Drive will be extended later by the periodic monitor.
+            self._log.warning("Executor queue full, retrying later")
+
+    def _extend_drive(self, drive):
+        if not self._update_block_info([drive]):
+            self._log.warning(
+                "Cannot update block info for drive %s, retrying later",
+                drive.name)
+            return
+
+        timeout = config.getfloat("thinp", "extend_timeout")
+        try:
+            with drive.monitor_lock(timeout):
```

Comment:
Can it happen that there are many drives with the lock taken on which this times out, resulting in delayed extensions of "good" drives, while there would be a reasonable way to prevent the starvation?  I guess realistically not at least in the sense that the previous version (without the lock) would be much better.  But asking to be sure.

Reply:
Theoretically it can happen, but is unlikely. We can avoid this if we allow up to one thread per vm for handling extends.